### PR TITLE
Add client side validation

### LINKED ISSUE #3149: [bug][production] Front-end URL validation not working as expected
## Summary
The logic used to validate URLs on the front-end does not seem to fire correctly. If an invalid URL is entered, the server sends a response "400/ BAD REQUEST", but this result is not displayed to the user. The field is not updated.

## Steps to reproduce
1. Go to https://osf.io/settings/
2. Click the "Social" tab
3. In the "Personal Site" box, type: `http://notaurl`
4. Click "submit"
5. Check the console for error messages. Refresh the page to confirm the value in the box was not saved to the server.

### Expected result
The user should not be allowed to submit an invalid value. Furthermore, the existing Knockout "url" validator should display an error message underneath the "personal site" box to warn of the problem and guide towards a solution.

### Actual result
Even though URL validation is turned on for this field, no error message is displayed. The "submit" button can be clicked, but no result is saved. Looking in the error console (which is not normal behavior for a user) gives the following:

`PUT https://osf.io/api/v1/settings/social/ 400 (BAD REQUEST)`

## Browsers tested
Same behavior in Firefox, Chrome

## References
- osf.io/website/static/js/profile.js : The "url" validator *appears* to be registered to the "personal" field in `SocialViewModel`
- osf.io/website/static/js/koHelpers.js : definition of the custom "url" validator.

### PR DESCRIPTION
**Pull request by billyhunt**
Fixes #3149 

The knockout front end URL validation does not work for observable arrays.  So, bad urls such as 

```
http://&lt;/script&gt;&lt;script&gt;alert(“Now I am doing bad stuff”)&lt;/script&gt;
```
and 
```
&lt;/script&gt;&lt;script&gt;alert(“Now I am doing bad stuff”)&lt;/script&gt;
```

generate either 400 or 500 errors on the server.

This update stops incorrect requests from being sent, giving relevant error messaging.  Note that I put the message at the top and the bottom.  The top message shows where the problem is, the bottom one kicks in if you press Save with invalid url information.  

screen shot 2015-10-07 at 9 17 29 am
Current

screen shot 2015-10-07 at 9 18 42 am
Proposed

### CODE REVIEW

**Comment by samanehsan** [comment]
I can't tell if this is just my localhost being weird, but can you enter a new employment or education position? I get an `Uncaught TypeError: this.hasValidWebsites is not a function`

File: website/static/js/profile.js [hunk 1]
```diff
@@ -484,8 +497,12 @@ var SocialViewModel = function(urls, modes) {
 
     self.addons = ko.observableArray();
 
+    self.profileWebsite = ko.observable('').extend({
+        ensureHttp: true,
```
Comment (samanehsan) [inline website/static/js/profile.js]: Please remove this extra comma

**Comment by samanehsan** [comment]
Pass complete @billyhunt :dart:

**Comment by samanehsan** [comment]
ACK! I forgot to mention, from talking with @caneruguz about UI/Style stuff: 

> The “please update your website before saving” text should not be below action buttons though
it should be either above or on the right side of Save button
it’s got enough padding to make a user not see by not scrolling enough

Could you move the error message at the bottom of the form above the cancel & save buttons?

File: website/static/js/profile.js [hunk 2]
```diff
@@ -20,7 +20,13 @@ var socialRules = {
     twitter: /twitter\.com\/(\w+)/i,
     linkedIn: /.*\/?(in\/.*|profile\/.*|pub\/.*)/i,
     impactStory: /impactstory\.org\/([\w\.-]+)/i,
-    github: /github\.com\/(\w+)/i
+    github: /github\.com\/(\w+)/i,
+    url: '^(https?:\\/\\/)?'+ // protocol
+            '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
+            '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
```
Comment (abought) [inline website/static/js/profile.js]: Will this also handle IPv6 addresses? (I hear they're all the rage)

**Comment by abought** [comment]
Even if this addresses the profile page, the URL validator is also used elsewhere- by itself it's not clear this PR will address the original ticket. (see, for example, http://localhost:5000/settings/applications/create/ : the client validator will allow the URL `http://----` through, but the server validator will reject the value)

**Comment by samanehsan** [comment]
I thought the ticket was just referring to the profile page, so thanks for pointing that out @abought. It could be useful to update that description with all of the places that have the URL validation issue.

**Comment by billyhunt** [comment]
Nice catch @abought.

**Comment by abought** [comment]
I haven't compiled a true master list of all pages that use the URL validator, but maybe this will help:

1. I *think* the URL validator is the one defined in `koHelpers.js`- https://github.com/abought/osf.io/blob/develop/website/static/js/koHelpers.js#L171

2. I see `.extend(... url: true` in these files:
- website/addons/forward/static/forwardConfig.js
- website/static/js/apiApplication.js
- website/static/js/profile.js

I'm not 100% sure if adding it as an extender as done in these pages is the right way to add the validator- on the one hand I copied existing syntax, but on the other hand the validator didn't work on existing pages. So there are a few possibilities for how to debug. :)

**Comment by sloria** [comment]
Closing this for now until it makes it back to the priority queue.

**Pull request closed without merging**